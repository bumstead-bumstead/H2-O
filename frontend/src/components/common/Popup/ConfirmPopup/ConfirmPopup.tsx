import { HTMLAttributes, MouseEventHandler, PropsWithChildren } from 'react';
import styled from '@emotion/styled';
import { CTACTypeButton, Popup } from '@/components/common';

export interface Props extends HTMLAttributes<HTMLDivElement> {
  hasCancelButton: boolean;
  cancelButtonLabel?: string;
  confirmButtonLabel: string;
  handleClickCancelButton: MouseEventHandler<HTMLElement>;
  handleClickConfirmButton: MouseEventHandler<HTMLButtonElement>;
}

/**
 * @example
 * ```tsx
 * ConfirmPopup hasCancelButton={true} cancelButtonLabel='취소' confirmButtonLabel='종료' handleClickCancelButton={() => setIsOpen(false)}
      handleClickConfirmButton={() => {window.open('https://www.naver.com');}}>
        내 차 만들기를 종료하시겠습니까? <br /> 지금 서비스 종료 시 저장되지 않습니다.
  </ConfirmPopup>
 * ```
 */

function ConfirmPopup({
  hasCancelButton,
  cancelButtonLabel = '취소',
  confirmButtonLabel,
  handleClickCancelButton,
  handleClickConfirmButton,
  children,
  ...restProps
}: PropsWithChildren<Props>) {
  return (
    <Popup handleClickDimmed={handleClickCancelButton} {...restProps}>
      <ConfirmPopupContainer data-testid='confirm-modal'>
        <ContentContainer>{children}</ContentContainer>
        <ButtonContainer>
          {hasCancelButton && (
            <CTACTypeButton size='small' color='gray' onClick={handleClickCancelButton}>
              {cancelButtonLabel}
            </CTACTypeButton>
          )}
          <CTACTypeButton size={hasCancelButton ? 'small' : 'large'} color='primary' onClick={handleClickConfirmButton}>
            {confirmButtonLabel}
          </CTACTypeButton>
        </ButtonContainer>
      </ConfirmPopupContainer>
    </Popup>
  );
}

export default ConfirmPopup;

const ConfirmPopupContainer = styled.div`
  ${({ theme }) => theme.flex.flexEndCol}
  width: 100%;
  height: 100%;
`;

const ContentContainer = styled.div`
  ${({ theme }) => theme.flex.flexCenterCol}
  ${({ theme }) => theme.typography.TextKRMedium14}
  color: ${({ theme }) => theme.colors.gray900};
  text-align: center;
  height: 100%;
`;

const ButtonContainer = styled.div`
  ${({ theme }) => theme.flex.flexCenterRow};
`;
