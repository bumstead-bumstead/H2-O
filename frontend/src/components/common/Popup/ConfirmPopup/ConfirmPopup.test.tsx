import ConfirmPopup from './ConfirmPopup';
import { fireEvent, render, screen } from '@/tests/test-util';

describe('Confirm Popup 컴포넌트 테스트', () => {
  beforeEach(() => {
    const portalRoot = global.document.createElement('div');
    portalRoot.setAttribute('id', 'portal-root');

    const body = global.document.querySelector('body');
    if (!body) throw new Error('portal-root의 body를 찾을 수 없습니다.');
    body.appendChild(portalRoot);
  });

  const props = {
    hasCancelButton: true,
    confirmButtonLabel: '확인',
    handleClickCancelButton: vi.fn(),
    handleClickConfirmButton: vi.fn(),
  };

  it('ConfirmPopup 컴포넌트가 팝업에 정상적으로 렌더링 된다.', () => {
    render(<ConfirmPopup {...props} />);

    expect(screen.getByTestId('dimmed')).toBeInTheDocument();
    expect(screen.getByRole('dialog')).toBeInTheDocument();
    expect(screen.getByTestId('confirm-modal')).toBeInTheDocument();
  });

  it('ConfirmPopup 컴포넌트의 children이 정상적으로 렌더링된다.', () => {
    render(<ConfirmPopup {...props}>Hello World</ConfirmPopup>);

    expect(screen.getByText('Hello World')).toBeInTheDocument();
  });

  it('컴포넌트의 프로퍼티 중 hasCancelButton이 true이면 버튼이 두 개 렌더링된다.', () => {
    render(<ConfirmPopup {...props} />);

    const buttons = screen.getAllByRole('button');
    expect(buttons).toHaveLength(2);
  });

  it('컴포넌트의 프로퍼티 중 hasCancelButton이 false이면 버튼이 한 개 렌더링된다.', () => {
    const newProps = { ...props, hasCancelButton: false };
    render(<ConfirmPopup {...newProps} />);

    const buttons = screen.getAllByRole('button');
    expect(buttons.length).toBe(1);
  });

  it('컴포넌트의 프로퍼티 중 cancelButtonLabel의 default값은 취소이다.', () => {
    render(<ConfirmPopup {...props} />);

    const buttons = screen.getAllByRole('button');
    expect(buttons[0].innerHTML).toBe('취소');
  });

  it('컴포넌트의 프로퍼티 중 cancelButtonLabel과 confirmButtonLabel을 통해서 버튼의 이름을 정할 수 있다.', () => {
    const newProps = { ...props, cancelButtonLabel: '종료', confirmButtonLabel: '확인' };
    render(<ConfirmPopup {...newProps} />);

    const buttons = screen.getAllByRole('button');
    expect(buttons[0].innerHTML).toBe('종료');
    expect(buttons[1].innerHTML).toBe('확인');
  });

  it('컴포넌트의 cancelButton을 클릭할 경우 handleClickCancelButton이 실행된다.', () => {
    const newProps = { ...props, handleClickCancelButton: vi.fn() };
    render(<ConfirmPopup {...newProps} />);

    const cancelButton = screen.getByText('취소');
    fireEvent.click(cancelButton);

    expect(newProps.handleClickCancelButton).toBeCalledTimes(1);
  });

  it('컴포넌트의 confirmButton을 클릭할 경우 handleClickConfirmButton이 실행된다.', () => {
    const newProps = { ...props, handleClickConfirmButton: vi.fn() };
    render(<ConfirmPopup {...newProps} />);

    const confirmButton = screen.getByText(props.confirmButtonLabel);
    fireEvent.click(confirmButton);

    expect(newProps.handleClickConfirmButton).toBeCalledTimes(1);
  });

  it('dimmed영역을 클릭할 경우 handleClickCancel이 실행된다.', () => {
    const newProps = { ...props, handleClickCancelButton: vi.fn() };
    render(<ConfirmPopup {...newProps} />);

    const dimmed = screen.getByTestId('dimmed');
    fireEvent.click(dimmed);

    expect(newProps.handleClickCancelButton).toBeCalledTimes(1);
  });

  it('ConfirmPopup 컴포넌트 스냅샷 테스트', () => {
    render(<ConfirmPopup {...props}>Snapshot</ConfirmPopup>);

    expect(screen.getByTestId('confirm-modal')).toMatchSnapshot();
  });
});
