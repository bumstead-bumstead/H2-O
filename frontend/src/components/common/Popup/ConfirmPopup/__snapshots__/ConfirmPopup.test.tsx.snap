// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Confirm Popup 컴포넌트 테스트 > ConfirmPopup 컴포넌트 스냅샷 테스트 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: end;
  -ms-flex-pack: end;
  -webkit-justify-content: flex-end;
  justify-content: flex-end;
  width: 100%;
  height: 100%;
}

.emotion-1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  font-family: Hyundai Sans Text KR,sans-serif;
  font-size: 14px;
  font-weight: 500;
  line-height: 22px;
  letter-spacing: -0.03em;
  color: #212122;
  text-align: center;
  height: 100%;
}

.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.emotion-3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  font-family: Hyundai Sans Head KR,sans-serif;
  font-size: 16px;
  font-weight: 500;
  line-height: 24px;
  letter-spacing: -0.03em;
  background-color: #C8CACC;
  color: #FFFFFF;
  width: 168px;
  height: 52px;
}

.emotion-3:active {
  background-color: #B6B8BB;
}

.emotion-4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  font-family: Hyundai Sans Head KR,sans-serif;
  font-size: 16px;
  font-weight: 500;
  line-height: 24px;
  letter-spacing: -0.03em;
  background-color: #00224A;
  color: #FFFFFF;
  width: 168px;
  height: 52px;
}

.emotion-4:active {
  background-color: #001731;
}

<div
  class="emotion-0"
  data-testid="confirm-modal"
>
  <div
    class="emotion-1"
  >
    Snapshot
  </div>
  <div
    class="emotion-2"
  >
    <button
      class="emotion-3"
      color="gray"
    >
      취소
    </button>
    <button
      class="emotion-4"
      color="primary"
    >
      확인
    </button>
  </div>
</div>
`;
